subsys/net/ip/net_pkt.c

```diff
@@ -1654,6 +1654,59 @@ void net_pkt_get_info(struct k_mem_slab **rx,
 	}
 }
 
+int net_pkt_get_src_addr(struct net_pkt *pkt, struct sockaddr *addr,
+			 socklen_t addrlen)
+{
+	struct net_context *ctx;
+	enum net_ip_protocol proto;
+	sa_family_t family;
+	u16_t port = 0;
+
+	if (!addr || !pkt) {
+		return -EINVAL;
+	}
+
+	ctx = pkt->context;
+	proto = net_context_get_ip_proto(ctx);
+	family = net_context_get_family(ctx);
```

Comment:
The pkt might not have context specified so this is error prone.
Family can be get using net_pkt_family(pkt) instead. For protocol, we could get that information from IP header, or we could store it to net_pkt header when packet is received. We have been trying to avoid increasing the size of net_pkt so first option could be done here.

So depending on family, the UDP/TCP information can be get like this
`proto = NET_IPV6_HDR(pkt)->nexthdr;`
or
`proto = NET_IPV4_HDR(pkt)->proto;`